shell/renderer/api/electron_api_context_bridge.cc

```diff
@@ -810,13 +857,233 @@ bool OverrideGlobalPropertyFromIsolatedWorld(
   }
 }
 
+// Determine if the current context is the main world context.
 bool IsCalledFromMainWorld(v8::Isolate* isolate) {
-  auto* render_frame = GetRenderFrame(isolate->GetCurrentContext()->Global());
-  CHECK(render_frame);
-  auto* frame = render_frame->GetWebFrame();
-  CHECK(frame);
-  v8::Local<v8::Context> main_context = frame->MainWorldScriptContext();
-  return isolate->GetCurrentContext() == main_context;
+  v8::Local<v8::Context> source_context = isolate->GetCurrentContext();
+  auto* ec = blink::ExecutionContext::From(source_context);
+  if (ec->IsWindow()) {
+    auto* render_frame = GetRenderFrame(source_context->Global());
+    CHECK(render_frame);
+    auto* frame = render_frame->GetWebFrame();
+    CHECK(frame);
+    v8::Local<v8::Context> main_context = frame->MainWorldScriptContext();
+    return source_context == main_context;
+  } else if (ec->IsShadowRealmGlobalScope()) {
+    return false;
+  } else if (ec->IsServiceWorkerGlobalScope()) {
+    return true;
+  } else {
+    NOTREACHED();
+  }
+}
+
+// Clones a value into the target context.
+v8::MaybeLocal<v8::Value> CloneValueToContext(
```

Comment:
Can you explain why we need this helper and can't just call `PassValueToOtherContext` directly.

* The early out for `source === destination` shouldn't matter and that case should be impossible, we should greedily clone if we ever end up in that unexpected state for safety anyway

`CloneValueToContext` doesn't imply it's "maybe" nature until you look at the return type, given the try/catch we probably want `MaybeCloneValueToContext`. But, I'd rather we just inline as this logic is single-use anyway.

Reply:
The explanation is that it was an artifact of a few refactors while I was writing the initial implementation. I looked it over again and agree it can be rewritten inline 👍